### ISSUE TITLE
Login/Logout flow

### ISSUE THREAD

**Issue by mistre833** [yelouafi/redux-saga#382]
Hi to all,
i've just started to use Saga today and i've implemented my login following this tutorial:

http://yelouafi.github.io/redux-saga/docs/advanced/NonBlockingCalls.html

Login works fine but how can handle LOGOUT ? This guide show how to prevent LOGOUT button pressed before the LOGIN cycle is completed.

When the user is logged in and for example him refresh the page, this fork wait again for START_LOGIN.

Moreover, usually LOGOUT is async like login, so we should use the same flow (logoutFlow) to handle LOGOUT ?

I've implemented LOGOUT in this way but, because i refresh the page the "first" fork will not be cancelled...

**Comment by kuy** [comment]
@mistre833 How about to separate login and logout flows and skip login if it's already authorized?

**Comment by mistre83** [comment]
Yes, this is what i have implemented. I have created two separate folow (watchLogin and watchLogout) where both start forking and with an indefinite while (while (true) ) i listen for actions and when this is fired, i fork again to call the API. In this way, i dont have problems.

Now, i have to read better the documentation and understand what redux-saga offeer (i think right now i have used just 10% of saga :) )

**Comment by mxstbr** [comment]
This is a really good showcase of a login flow with redux-saga: https://github.com/sotojuan/saga-login-flow

**Comment by mistre83** [comment]
Thanks a lot!!! I'm looking it :)

**Comment by mistre83** [comment]
I'm trying to write some test with this flow, and my login watchers is this:

```
export function* watchLogin() {
    while (true) {
        const user = yield take(actions.LOGIN_USER);
        const { username, password } = user.payload;
        yield fork(loginCall, username, password);
    }
}
```

Right now i've wrote this test but.... it fail :(

```
it('watchLogin Saga', (done) => {
        const generator = watchLogin();
        expect(generator.next().value).toEqual(take(actions.LOGIN_USER));
        generator.next({'payload': {'username': 'admin', 'password': 'admin'}});
        expect(generator.next({'payload': {'username': 'admin', 'password': 'admin'}}).value)
            .toEqual(fork(loginCall, 'admin', 'admin'));
        expect(generator.next().value).toEqual(take(actions.LOGIN_USER));
        done();
    });
```

The first expect works (sure, is simple :D) the second works too.. but now... i have problems with third and fourth step:

- Third step: should fork with username and password provided in the previous step
- Fourth step: should go back to the first (no warried about if login is fine or not, in loginCall i catch the exception and dispatch action for LOGIN_OK and LOGIN_FAIL).

Can you help me to write this simple test? I'm trying to understand how to write it looking on examples but is a little difficult... i'm new to all.. :D Redux, React, Saga, Generator... :)

**Comment by kuy** [comment]
@mistre83 You call `next()` with payload twice in your test code. Please remove the line `generator.next({'payload': {'username': 'admin', 'password': 'admin'}});` like this.

```
it('watchLogin Saga', (done) => {
        const generator = watchLogin();
        expect(generator.next().value).toEqual(take(actions.LOGIN_USER));
        expect(generator.next({'payload': {'username': 'admin', 'password': 'admin'}}).value)
            .toEqual(fork(loginCall, 'admin', 'admin'));
        expect(generator.next().value).toEqual(take(actions.LOGIN_USER));
        done();
    });
```

**Comment by mistre833** [comment]
Thanks!! tomorrow i will test it again :)

**Comment by yelouafi** [comment]
@mistre833  please note that usage questions should be posted on SO

**Comment by mistre83** [comment]
Yes... sorry for this i think we can close this.
